compiler/rustc_ast/src/mut_visit.rs

```diff
@@ -1649,7 +1653,7 @@ impl DummyAstNode for Ty {
     fn dummy() -> Self {
         Ty {
             id: DUMMY_NODE_ID,
-            kind: TyKind::Err,
+            kind: TyKind::Tup(ThinVec::new()),
```

Comment:
An explicit dummy TyKind could help ensure we don't accidentally keep processing a `()` in hir. Not sure how realistic that worry is

Reply:
We already use `Wild` for pattern dummies. I deliberately don't want to add a dummy variant because then we need to handle it in a lot of places. In #120586 I'm advocating that `ExprKind::Dummy` not be added.

Reply:
And an empty `Tup` is already used for some `ExprKind` dummies.

Reply:
What oli is saying is exactly what I also wanted to say in the `ExprKind::Err(ErrorGuaranteed)` PR, having `Dummy` signals the intend better and it's safer to you since it leads to a `bug!()` when you attempt to lower it.

Reply:
For example, see https://github.com/rust-lang/rust/pull/120586#discussion_r1482296982. Using `Dummy` there isn't great but it's it feels more correct than `Tup([])`. `ExprKind::Err` over there could be an option if we delayed a bug but not sure if that's a good idea.

Reply:
We could make `Err` take an `Option<ErrorGuaranteed>` and always unwrap it 😆